exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,434 @@
+    /// Unable to parse artifact identifier or specified artifact has non-empty spec.
+    IncorrectArtifactId = 0,
+    /// Checked java exception is occurred
+    JavaException = 1,
+    /// Unspecified error
+    UnspecifiedError = 2,
+    /// Not supported operation
+    NotSupportedOperation = 3,
+}
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceStateHashes")]
+struct ServiceStateHashes {
+    instance_id: u32,
+    state_hashes: Vec<Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceRuntimeStateHashes")]
+struct ServiceRuntimeStateHashes {
+    runtime_state_hashes: Vec<Vec<u8>>,
+    service_state_hashes: Vec<ServiceStateHashes>,
+}
+
+impl JavaRuntimeProxy {
+    /// Runtime Identifier
+    pub const RUNTIME_ID: RuntimeIdentifier = RuntimeIdentifier::Java;
+
+    /// Creates new `JavaRuntimeProxy` for given `ServiceRuntimeAdapter` object
+    pub fn new(executor: Executor, adapter: GlobalRef) -> Self {
+        JavaRuntimeProxy {
+            exec: executor,
+            runtime_adapter: adapter,
+        }
+    }
+
+    fn parse_artifact(&self, artifact: &ArtifactId) -> Result<JavaArtifactId, ExecutionError> {
+        if artifact.runtime_id != Self::RUNTIME_ID as u32 {
+            Err(Error::IncorrectArtifactId.into())
+        } else {
+            artifact
+                .name
+                .parse()
+                .map_err(|_| Error::IncorrectArtifactId.into())
+        }
+    }
+
+    fn parse_jni<T>(res: JniResult<T>) -> Result<T, ExecutionError> {
+        res.map_err(|err| match err.0 {
+            JniErrorKind::JavaException => Error::JavaException.into(),
+            _ => Error::UnspecifiedError.into(),
+        })
+    }
+
+    /// If the current node is a validator, returns `Some(validator_id)`, for other nodes return `None`.
+    fn validator_id(snapshot: &Snapshot, service_keypair: &(PublicKey, SecretKey)) -> Option<ValidatorId> {
+        CoreSchema::new(snapshot)
+            .actual_configuration()
+            .validator_keys
+            .iter()
+            .position(|validator| service_keypair.0 == validator.service_key)
+            .map(|id| ValidatorId(id as u16))
+    }
+
+    /// If the current node is a validator, return its identifier, for other nodes return `default`.
+    fn validator_id_or(
+        snapshot: &Snapshot,
+        service_keypair: &(PublicKey, SecretKey),
+        default: i32
+    ) -> i32 {
+        Self::validator_id(snapshot, service_keypair)
+            .map_or(default, |id| i32::from(id.0))
+    }
+}
+
+impl fmt::Debug for JavaRuntimeProxy {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "JavaRuntimeProxy()")
+    }
+}
+
+impl From<Error> for ExecutionError {
+    fn from(value: Error) -> ExecutionError {
+        ExecutionError::new(
+            ErrorKind::runtime(value as u8),
+            format!("{:?}", value.clone()),
```

Comment:
This isn't a valid description, is it?

Reply:
fixed in https://github.com/exonum/exonum-java-binding/pull/1129/commits/17a658560ef4f58bcc3359d4a8251b90002da62f